aiomcache/client.py

```diff
@@ -415,8 +444,357 @@ def version(self, conn):
     def flush_all(self, conn):
         """Its effect is to invalidate all existing items immediately"""
         command = b'flush_all\r\n'
-        response = yield from self._execute_simple_command(
-            conn, command)
+        response = yield from self._execute_simple_command(conn, command)
 
         if const.OK != response:
             raise ClientException('Memcached flush_all failed', response)
+
+
+class BinaryClient(ClientBase):
+
+    # binary protocol description:
+    # https://github.com/memcached/memcached/blob/master/doc/protocol-binary.xml
+
+    @staticmethod
+    def _make_command(opcode, key=b'', data=b'', extra=b'', data_type=0,
+                      status=0, opaque=0, cas_token=0):
+        return struct.pack(
+            '>BBHBBHLLQ', 0x80, int(opcode), len(key), len(extra), data_type,
+            status, len(extra) + len(key) + len(data), opaque, cas_token) + \
+            extra + key + data
+
+    @asyncio.coroutine
+    def _send_command(self, conn, raw_command):
+        conn.writer.write(raw_command)
+        yield from conn.writer.drain()
+
+    @asyncio.coroutine
+    def _receive_response(self, conn):
+        header = yield from conn.reader.read(24)
+        opcode, key_len, extra_len, data_type, status, body_len, _, cas = \
+            struct.unpack('>xBHBBHLLQ', header)
+
+        if extra_len:
+            extra = yield from conn.reader.read(extra_len)
+        else:
+            extra = b''
+
+        if key_len:
+            key = yield from conn.reader.read(key_len)
+        else:
+            key = b''
+
+        data_len = body_len - key_len - extra_len
+        if data_len:
+            data = yield from conn.reader.read(data_len)
+        else:
+            data = b''
+
+        return (opcode, key, data, extra, status, cas)
+
+    @asyncio.coroutine
+    def _execute_simple_command(self, conn, raw_command):
+        # send a command and return the response status
+        yield from self._send_command(conn, raw_command)
+        return (yield from self._receive_response(conn))[4]
+
+    @asyncio.coroutine
+    def _multi_get(self, conn, *keys, with_cas=False, with_flags=False):
+        if not keys:
+            return {}, {}, {}
+
+        if len(set(keys)) != len(keys):
+            raise ClientException('duplicate keys passed to multi_get')
+        [self._validate_key(key) for key in keys]
+
+        # send GETQ for all keys except the last one
+        # this allows the server to skip sending responses for missing keys
+        commands = b''.join(self._make_command(const.Opcode.GETKQ, k)
+                            for k in keys[:-1])
+        yield from self._send_command(conn, commands)
```

Comment:
can we pack this two calls into one? I mean send both commands with one write to socket?